lib/SILOptimizer/Mandatory/OSLogOptimization.cpp

```diff
@@ -428,91 +434,178 @@ static SILValue emitCodeForSymbolicValue(SymbolicValue symVal,
+      if (potentialEndUsers.count(inst)) {
+        lastUserOfValue = inst;
+        break;
       }
     }
+    assert(lastUserOfValue);
+    endUsers.push_back(lastUserOfValue);
+  }
+}
+
+/// Return the first use of \c val that is not a retain or release or a destory
+/// instruction. Otherwise return nullptr.
+static Operand *getFirstNonARCUse(SILValue val) {
+  for (Operand *use : val->getUses()) {
+    SILInstruction *user = use->getUser();
+    if (isa<ReleaseValueInst>(user) || isa<RetainValueInst>(user) ||
+        isa<DestroyValueInst>(user)) {
+      continue;
+    }
+    return use;
   }
+  return nullptr;
 }
 
-/// Emit instructions to destroy the folded value at the end of its use, if
-/// required. Since this pass folds only integers or strings and since the
-/// former is a trivial type, we only have to destroy strings that are folded.
-/// For strings, a release_value (or a destory_value instruction in ownership
-/// SIL) has to be emitted if it is not already present.
+/// Replace all uses of \c originalVal by \c foldedVal and adjust lifetimes of
+/// original and folded values by emitting required destory/release instructions
+/// at the right places. Note that this function does not remove any
+/// instruction. \param originalVal the SIL value that is replaced. \param
+/// foldedVal the SIL value that replaces the \c originalVal. \param
+/// endOfUseInsts instructions marking the end of use of the \c foldedVal \param
+/// fun the SIL function containing the \c foldedVal and \c originalVal
 static void
-destroyFoldedValueAtEndOfUse(SILValue foldedVal, SILValue originalVal,
-                             ArrayRef<SILInstruction *> endOfUseInsts,
-                             SILFunction *fun) {
-  // Folded value should have either trivial or owned ownership as it is an
-  // integer or string constant.
-  assert(foldedVal.getOwnershipKind() == ValueOwnershipKind::Any ||
-         foldedVal.getOwnershipKind() == ValueOwnershipKind::Owned);
-
-  // If the ownership kinds of folded and original values are both either
-  // owned or trivial, there is nothing to do.
-  if (foldedVal.getOwnershipKind() == originalVal.getOwnershipKind()) {
+replaceAllUsesAndFixLifetimes(SILValue foldedVal, SILValue originalVal,
+                              ArrayRef<SILInstruction *> endOfUseInsts,
+                              SILFunction *fun) {
+  if (originalVal->getType().isTrivial(*fun)) {
+    assert(foldedVal->getType().isTrivial(*fun));
+    originalVal->replaceAllUsesWith(foldedVal);
     return;
   }
-  assert(originalVal.getOwnershipKind() == ValueOwnershipKind::Guaranteed);
+  assert(!foldedVal->getType().isTrivial(*fun));
+
+  // A helper for releasing values in OSSA and non-OSSA.
+  auto releaseValue = [&](SILBuilder &builder, SILLocation loc, SILValue val) {
+    if (fun->hasOwnership()) {
+      builder.createDestroyValue(loc, val);
+    } else {
+      builder.createReleaseValue(loc, val, builder.getDefaultAtomicity());
```

Comment:
You can use emitDestroyValueOperation here.

Reply:
Sure.